File: jslua/lexer.js
// Analizador Léxico de Lua

function get_fail_lines (str, pos) {
  var startpos = pos;
  var endpos = pos;
  var inpos;
  while (startpos > 0) {
    if (str[startpos] == '\n') {
      startpos = pos + 1;
      break;
    }
    startpos--;
  }
  while (endpos > 0) {
    if (str[endpos] == '\n') {
      endpos = pos;
      break;
    }
    endpos--;
  }
  var inpos = (pos+1)-startpos;
  var pointer = "";
  for (var i = 0; i < inpos; i++) {
    pointer += " ";
  };
  pointer += "^";
  return [str.slice(startpos, endpos), pointer];
}

function token (type, val, pos) {
  return {
    type: type,
    match: val,
    pos: pos
  };
}

function startsWith(str, prefix) {
  return str.slice(0, prefix.length) == prefix;
}

function startsWithAny (str, prefixes) {
  for (var i = 0; i < prefixes.length; i++) {
    var prefix = prefixes[i];
    if (startsWith(str, prefix)) return prefix;
  };
  return null;
}

function tokenize (input) {
  var left = input;
  var tokens = [];
  function fail (msg) {
    var pos = (input.length-left.length);
    var lines = get_fail_lines(input, pos);
    var str = "Lexer Error: Token Irreconocible.\nat: " + pos;
    str += '\n' + lines[0] + '\n' + lines[1];
    console.log(input);
    console.log(tokens);
    throw new Error(str);
  }
  function consume (len) { left = left.slice(len); }
  function getpos () { return input.length - left.length; }
  function consume_space () {
    var regex = /[\ \t]*/; // No se consume las nuevas líneas.
    var match = regex.exec(left);
    if (match) { consume(match[0].length); }
  }
  function mRegex (rgx, type) {
    var regex = RegExp('^' + rgx.source);
    var match = regex.exec(left);
    if (match == null) return false;
    var tkn = token(type, match[0], getpos());
    tokens.push(tkn);
    consume(tkn.match.length);
    return true;
  }
  function mAny (words, type) {
    var match = startsWithAny(left, words);
    if (match == null) return false;
    var tkn = token(type, match, getpos());
    tokens.push(tkn);
    consume(tkn.match.length);
    return true;
  }

  var numrgx = /[0-9]+(\.[0-9]+)?([eE][-+]?[0-9]+)?/;
  var varrgx = /[a-z][a-zA-Z0-9]*/;
  var kws = ["do", "else", "elseif", "end", "for", "function", "if", "in",
             "repeat", "then", "until", "while"];
  var stmts = ["local", "break", "return"];
  var cons = ["true", "false", "nil", "..."];
  var ops = ["+", "-", "*", "/", "%", "^", "<", ">", "==", "<=", ">=", "~=",
             "and", "or", "not", "..", "#"];
  var grammar = ["[", "]", "{", "}", "(", ")", ".", ",", "="];

  var count = 100;
  while (left.length > 0 && count > 0) {
    count--;
    consume_space();
    if (left[0] == "\n" || left[0] == ";") {
      tokens.push(token("nl", left[0]));
      consume(1);
      continue;
    }
    if (mAny(stmts, "stmt")) continue;
    if (mAny(cons, "const")) continue;
    if (mAny(kws, "kw")) continue;
    if (mAny(ops, "op")) continue;
    if (mRegex(numrgx, "num")) continue;
    if (mRegex(varrgx, "var")) continue;
    if (mAny(grammar, "gmr")) continue;
    fail("Lexer Error: Token Irreconocible");
  }
  return tokens;
}

exports.tokenize = tokenize;

// Hay un error con este lexer, "ifyolo" devuelve "if" separado de "yolo",
// en vez de considerarlo todo como un solo

/*
var lexer = require("./lexer.js"); var T = lexer.tokenize;
lexer.tokenize("3 4 5")
*/

File: rubylua/lexer.rb
require_relative 'base'

class Token
  attr_reader :type, :match, :pos
  def initialize type, match = nil
    @type = type
    @match = match
  end
  def set_pos pos
    @pos = pos
  end
  def length
    @match.length
  end
  def to_s
    "<Token-#{@type}: #{@match}>"
  end
end

class LexerState
  attr_reader :text, :token, :tokens, :pos
  def initialize text
    @text = text
    @token = nil
    @tokens = []
    @pos = 0
  end
  def consume len
    @pos += len
    @text = @text[len .. -1]
  end
  def consume_space
    result = /\s*/.match @text
    consume result[0].length if result
  end
  def try_match patt, type
    match = nil
    case patt
    when Array
      longest = nil
      patt.each do |p|
        if @text.start_with? p then
          longest = p if is_longer? p, longest
        end
      end
      match = Token.new type, longest if longest
    when Regexp
      result = patt.match @text
      if result.nil? or
        result.length == 0 or
        result.begin(0) > 0 or
        result[0].length == 0
      then return end
      match = Token.new type, result[0]
    end
    @token = match if is_longer? match, @token
  end
  def use_token
    raise "Match is Nil!" if @token.nil?
    @token.set_pos @pos
    @tokens.push @token
    consume @token.length
    @token = nil
  end

  ## Helpers
  private
  def is_longer? newer, older
    (not newer.nil?) and (older.nil? or newer.length > older.length)
  end
end

def tokenize input
  text = if input.class == String
    then input else input.text end
  state = LexerState.new text

  ### Constantes para comparar
  numrgx = /[0-9]+(\.[0-9]+)?([eE][-+]?[0-9]+)?/
  varrgx = /[a-z][a-zA-Z0-9]*/
  strrgx = /\"([^\\^\"]|\\.)*\"/
  kws = %w(do elseif else end for function if in repeat then until while local break return)
  cons = %w(true false nil ...)
  ops = %w(+ - * / % ^ == <= >= < > ~= and or not .. #)
  grammar = %w(; [ ] { } ( ) . , =)
  # %w es sintaxis especial para arrays de palabras

  ### Procedimiento
  while state.text.length > 0
    state.consume_space
    state.try_match(kws, :kw)
    state.try_match(ops, :op)
    state.try_match(cons, :const)
    state.try_match(numrgx, :num)
    state.try_match(varrgx, :var)
    state.try_match(strrgx, :str)
    state.try_match(grammar, :gmr)
    raise "Unrecognized token. #{text}" if state.token.nil?
    state.use_token
  end
  return state.tokens
end